File: ios/Features/System/SystemBridge.swift
import Flutter

class SystemBridge {
  static var shared: SystemBridge?

  private let api: SystemFlutterApi

  init(binaryMessenger: FlutterBinaryMessenger) {
    api = SystemFlutterApi(binaryMessenger: binaryMessenger)
  }

  static func logDebug(_ message: String, logger: String) {
    shared?.log(message, logger: logger, level: "DEBUG")
  }

  static func logInfo(_ message: String, logger: String) {
    shared?.log(message, logger: logger, level: "INFO")
  }

  static func logError(_ message: String, logger: String) {
    shared?.log(message, logger: logger, level: "ERROR")
  }

  func log(_ message: String, logger: String, level: String = "INFO", error: String? = nil) {
    api.log(
      entry: LogData(level: level, loggerName: logger, message: message, error: error)
    ) { _ in }
  }

  func notifyAppResumed() {
    api.onAppResumed { _ in }
  }

  func notifyBackgroundFetch(completion: @escaping () -> Void) {
    api.onBackgroundFetch { _ in completion() }
  }
}

File: lib/bridge/system_bridge.dart
// coverage:ignore-file
import 'package:logging/logging.dart';

import '../config/dependencies.dart';
import '../data/services/local/storage/config_store_service.dart';
import '../domain/sync/sync_manager.dart';
import '../pigeon/system.g.dart';

class SystemBridge implements SystemFlutterApi {
  SystemBridge() {
    SystemFlutterApi.setUp(this);
  }

  static final _levelByName = {for (final l in Level.LEVELS) l.name: l};

  @override
  void log(LogData entry) {
    final level = _levelByName[entry.level] ?? Level.INFO;
    Logger(entry.loggerName).log(level, entry.message, entry.error);
  }

  @override
  Future<void> onAppResumed() async {
    await dependencies.get<ConfigStoreService>().reload();
    await SyncManager.instance.synchronize(withFinalRefresh: true);
  }

  @override
  Future<void> onBackgroundFetch() async {
    await SyncManager.instance.synchronize(withFinalRefresh: true);
  }
}
